File: langs/ocaml/minijson.ml
(* Tiny JSON encoder/decoder. No opam packages. *)

type value =
  | JNull
  | JBool of bool
  | JInt of int64
  | JStr of string
  | JArr of value list
  | JObj of (string * value) list

let arg_int args i =
  match List.nth args i with
  | JInt n -> n
  | JStr s -> Int64.of_string s
  | _ -> failwith "expected int arg"

let arg_str args i =
  match List.nth args i with
  | JStr s -> s
  | _ -> failwith "expected string arg"

let obj_val row key =
  match row with
  | JObj pairs ->
      let rec go = function
        | [] -> JNull
        | (k, v) :: rest -> if k = key then v else go rest
      in
      go pairs
  | _ -> JNull

let obj_str row key =
  match obj_val row key with
  | JStr s -> s
  | _ -> ""

let maybe_int = function
  | None -> JNull
  | Some n -> JInt n

let maybe_str = function
  | None -> JNull
  | Some s -> JStr s

let hex_digit c =
  match c with
  | '0' .. '9' -> Char.code c - Char.code '0'
  | 'a' .. 'f' -> 10 + Char.code c - Char.code 'a'
  | 'A' .. 'F' -> 10 + Char.code c - Char.code 'A'
  | _ -> failwith "bad hex"

let read_hex4 a b c d =
  ((hex_digit a * 16 + hex_digit b) * 16 + hex_digit c) * 16 + hex_digit d

let unescape = function
  | '"' -> '"'
  | '\\' -> '\\'
  | '/' -> '/'
  | 'b' -> '\b'
  | 'f' -> '\012'
  | 'n' -> '\n'
  | 'r' -> '\r'
  | 't' -> '\t'
  | c -> c

let rec skip s i =
  if i >= String.length s then i
  else
    match s.[i] with
    | ' ' | '\t' | '\n' | '\r' -> skip s (i + 1)
    | _ -> i

let rec parse_string s i acc =
  if i >= String.length s then failwith "unterminated string"
  else
    match s.[i] with
    | '"' -> (Buffer.contents acc, i + 1)
    | '\\' when i + 1 < String.length s ->
        if s.[i + 1] = 'u' && i + 5 < String.length s then (
          let code = read_hex4 s.[i + 2] s.[i + 3] s.[i + 4] s.[i + 5] in
          Buffer.add_utf_8_uchar acc (Uchar.of_int code);
          parse_string s (i + 6) acc)
        else (
          Buffer.add_char acc (unescape s.[i + 1]);
          parse_string s (i + 2) acc)
    | c ->
        Buffer.add_char acc c;
        parse_string s (i + 1) acc

let rec skip_digits s i =
  if i < String.length s && s.[i] >= '0' && s.[i] <= '9' then skip_digits s (i + 1)
  else i

let skip_exp s i =
  if i < String.length s && (s.[i] = 'e' || s.[i] = 'E') then
    let i = i + 1 in
    let i =
      if i < String.length s && (s.[i] = '+' || s.[i] = '-') then i + 1 else i
    in
    skip_digits s i
  else if i < String.length s && s.[i] = '.' then skip_digits s (i + 1)
  else i

let parse_unsigned s i =
  let n = String.length s in
  if i >= n || s.[i] < '0' || s.[i] > '9' then failwith "bad json number"
  else
    let rec go i acc =
      if i < n && s.[i] >= '0' && s.[i] <= '9' then
        let d = Int64.of_int (Char.code s.[i] - Char.code '0') in
        go (i + 1) (Int64.add (Int64.mul acc 10L) d)
      else (acc, skip_exp s i)
    in
    go i 0L

let parse_number s i =
  if i < String.length s && s.[i] = '-' then
    let n, rest = parse_unsigned s (i + 1) in
    (JInt (Int64.neg n), rest)
  else
    let n, rest = parse_unsigned s i in
    (JInt n, rest)

let rec parse_value s i =
  let i = skip s i in
  if i >= String.length s then failwith "bad json"
  else
    match s.[i] with
    | 'n'
      when i + 3 < String.length s
           && s.[i + 1] = 'u'
           && s.[i + 2] = 'l'
           && s.[i + 3] = 'l' ->
        (JNull, i + 4)
    | 't'
      when i + 3 < String.length s
           && s.[i + 1] = 'r'
           && s.[i + 2] = 'u'
           && s.[i + 3] = 'e' ->
        (JBool true, i + 4)
    | 'f'
      when i + 4 < String.length s
           && s.[i + 1] = 'a'
           && s.[i + 2] = 'l'
           && s.[i + 3] = 's'
           && s.[i + 4] = 'e' ->
        (JBool false, i + 5)
    | '[' -> parse_array s (skip s (i + 1)) []
    | '{' -> parse_object s (skip s (i + 1)) []
    | '"' ->
        let str, rest = parse_string s (i + 1) (Buffer.create 16) in
        (JStr str, rest)
    | c when c = '-' || (c >= '0' && c <= '9') -> parse_number s i
    | _ -> failwith "bad json"

and parse_array s i acc =
  let i = skip s i in
  if i < String.length s && s.[i] = ']' then (JArr (List.rev acc), i + 1)
  else
    let v, rest0 = parse_value s i in
    let rest = skip s rest0 in
    if rest < String.length s && s.[rest] = ',' then
      parse_array s (skip s (rest + 1)) (v :: acc)
    else if rest < String.length s && s.[rest] = ']' then
      (JArr (List.rev (v :: acc)), rest + 1)
    else failwith "bad json array"

and parse_object s i acc =
  let i = skip s i in
  if i < String.length s && s.[i] = '}' then (JObj (List.rev acc), i + 1)
  else if i < String.length s && s.[i] = '"' then
    let key, rest1 = parse_string s (i + 1) (Buffer.create 16) in
    let rest2 = skip s rest1 in
    if rest2 >= String.length s || s.[rest2] <> ':' then failwith "bad json object"
    else
      let v, rest4 = parse_value s (skip s (rest2 + 1)) in
      let rest5 = skip s rest4 in
      if rest5 < String.length s && s.[rest5] = ',' then
        parse_object s (skip s (rest5 + 1)) ((key, v) :: acc)
      else if rest5 < String.length s && s.[rest5] = '}' then
        (JObj (List.rev ((key, v) :: acc)), rest5 + 1)
      else failwith "bad json object"
  else failwith "bad json object"

let decode input =
  let v, rest = parse_value input 0 in
  if skip input rest <> String.length input then failwith "trailing json" else v

let rec encode = function
  | JNull -> "null"
  | JBool true -> "true"
  | JBool false -> "false"
  | JInt n -> Int64.to_string n
  | JStr s ->
      let buf = Buffer.create (String.length s + 2) in
      Buffer.add_char buf '"';
      String.iter
        (fun c ->
          match c with
          | '"' -> Buffer.add_string buf "\\\""
          | '\\' -> Buffer.add_string buf "\\\\"
          | '\b' -> Buffer.add_string buf "\\b"
          | '\012' -> Buffer.add_string buf "\\f"
          | '\n' -> Buffer.add_string buf "\\n"
          | '\r' -> Buffer.add_string buf "\\r"
          | '\t' -> Buffer.add_string buf "\\t"
          | _ when Char.code c < 32 ->
              Buffer.add_string buf (Printf.sprintf "\\u%04x" (Char.code c))
          | _ -> Buffer.add_char buf c)
        s;
      Buffer.add_char buf '"';
      Buffer.contents buf
  | JArr xs -> "[" ^ String.concat "," (List.map encode xs) ^ "]"
  | JObj pairs ->
      "{"
      ^ String.concat ","
          (List.map (fun (k, v) -> encode (JStr k) ^ ":" ^ encode v) pairs)
      ^ "}"

File: langs/haskell/MiniJson.hs
-- Tiny JSON encoder/decoder. No cabal packages.
module MiniJson
  ( Value (..)
  , decode
  , encode
  , argInt
  , argStr
  , objVal
  , objStr
  )
where

import Data.Char (chr, isDigit, ord)
import Data.Int (Int64)
import Numeric (showHex)

data Value
  = JNull
  | JBool Bool
  | JInt Integer
  | JStr String
  | JArr [Value]
  | JObj [(String, Value)]
  deriving (Eq, Show)

argInt :: [Value] -> Int -> Int64
argInt args i =
  case args !! i of
    JInt n -> fromInteger n
    JStr s -> fromInteger (read s)
    _ -> error "expected int arg"

argStr :: [Value] -> Int -> String
argStr args i =
  case args !! i of
    JStr s -> s
    _ -> error "expected string arg"

objVal :: Value -> String -> Value
objVal (JObj pairs) key = go pairs
  where
    go [] = JNull
    go ((k, v) : rest) = if k == key then v else go rest
objVal _ _ = JNull

objStr :: Value -> String -> String
objStr row key =
  case objVal row key of
    JStr s -> s
    _ -> ""

decode :: String -> Value
decode input =
  case value (skip input) of
    (val, rest)
      | null (skip rest) -> val
      | otherwise -> error "trailing json"

encode :: Value -> String
encode JNull = "null"
encode (JBool True) = "true"
encode (JBool False) = "false"
encode (JInt n) = show n
encode (JStr s) = '"' : escape s ++ "\""
encode (JArr xs) = '[' : intercalate "," (map encode xs) ++ "]"
encode (JObj pairs) =
  '{' : intercalate "," [encode (JStr k) ++ ":" ++ encode v | (k, v) <- pairs] ++ "}"

intercalate :: [a] -> [[a]] -> [a]
intercalate _ [] = []
intercalate sep (x : xs) = x ++ concatMap (sep ++) xs

escape :: String -> String
escape = concatMap esc
  where
    esc c
      | c == '"' = "\\\""
      | c == '\\' = "\\\\"
      | c == '\b' = "\\b"
      | c == '\f' = "\\f"
      | c == '\n' = "\\n"
      | c == '\r' = "\\r"
      | c == '\t' = "\\t"
      | ord c < 32 = "\\u" ++ pad4 (showHex (ord c) "")
      | otherwise = [c]
    pad4 hex = replicate (4 - length hex) '0' ++ hex

skip :: String -> String
skip (' ' : xs) = skip xs
skip ('\t' : xs) = skip xs
skip ('\n' : xs) = skip xs
skip ('\r' : xs) = skip xs
skip xs = xs

value :: String -> (Value, String)
value ('n' : 'u' : 'l' : 'l' : xs) = (JNull, xs)
value ('t' : 'r' : 'u' : 'e' : xs) = (JBool True, xs)
value ('f' : 'a' : 'l' : 's' : 'e' : xs) = (JBool False, xs)
value ('[' : xs) = parseArray (skip xs) []
value ('{' : xs) = parseObject (skip xs) []
value ('"' : xs) =
  let (s, rest) = parseString xs []
   in (JStr s, rest)
value xs@(c : _)
  | c == '-' || isDigit c = parseNumber xs
value _ = error "bad json"

parseArray :: String -> [Value] -> (Value, String)
parseArray (']' : xs) acc = (JArr (reverse acc), xs)
parseArray xs acc =
  let (val, rest0) = value (skip xs)
      rest = skip rest0
   in case rest of
        (',' : rest1) -> parseArray (skip rest1) (val : acc)
        (']' : rest1) -> (JArr (reverse (val : acc)), rest1)
        _ -> error "bad json array"

parseObject :: String -> [(String, Value)] -> (Value, String)
parseObject ('}' : xs) acc = (JObj (reverse acc), xs)
parseObject ('"' : xs) acc =
  let (key, rest1) = parseString xs []
      rest2 = skip rest1
   in case rest2 of
        (':' : rest3) ->
          let (val, rest4) = value (skip rest3)
              rest5 = skip rest4
           in case rest5 of
                (',' : rest6) -> parseObject (skip rest6) ((key, val) : acc)
                ('}' : rest6) -> (JObj (reverse ((key, val) : acc)), rest6)
                _ -> error "bad json object"
        _ -> error "bad json object"
parseObject _ _ = error "bad json object"

parseString :: String -> String -> (String, String)
parseString ('"' : xs) acc = (reverse acc, xs)
parseString ('\\' : 'u' : a : b : c : d : xs) acc =
  parseString xs (chr (readHex4 [a, b, c, d]) : acc)
parseString ('\\' : c : xs) acc = parseString xs (unescape c : acc)
parseString (c : xs) acc = parseString xs (c : acc)
parseString [] _ = error "unterminated string"

unescape :: Char -> Char
unescape '"' = '"'
unescape '\\' = '\\'
unescape '/' = '/'
unescape 'b' = '\b'
unescape 'f' = '\f'
unescape 'n' = '\n'
unescape 'r' = '\r'
unescape 't' = '\t'
unescape c = c

readHex4 :: String -> Int
readHex4 = go 0
  where
    go n [] = n
    go n (c : cs) = go (n * 16 + hexDigit c) cs

hexDigit :: Char -> Int
hexDigit c
  | c >= '0' && c <= '9' = ord c - ord '0'
  | c >= 'a' && c <= 'f' = 10 + ord c - ord 'a'
  | c >= 'A' && c <= 'F' = 10 + ord c - ord 'A'
  | otherwise = error "bad hex"

parseNumber :: String -> (Value, String)
parseNumber ('-' : xs) =
  case parseUnsigned xs of
    (n, rest) -> (JInt (-n), rest)
parseNumber xs =
  case parseUnsigned xs of
    (n, rest) -> (JInt n, rest)

parseUnsigned :: String -> (Integer, String)
parseUnsigned xs =
  let (digits, rest) = span isDigit xs
   in if null digits
        then error "bad json number"
        else (read digits, skipExp rest)

skipExp :: String -> String
skipExp (c : xs)
  | c == 'e' || c == 'E' =
      case xs of
        (sign : rest)
          | sign == '+' || sign == '-' -> snd (span isDigit rest)
        _ -> snd (span isDigit xs)
skipExp ('.' : xs) = snd (span isDigit xs)
skipExp xs = xs
